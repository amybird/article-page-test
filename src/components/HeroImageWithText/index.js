import React, { useState, useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import ResponsiveImage from '../ResponsiveImage';
import {
  HeroImageWrap, ImageWrap, Overlay, TextOverlay
} from './styles';

const HeroImageWithText = ({ data }) => {
  const [isScrollTop, setIsScrollTop] = useState(false);
  const heroImageWrap = useRef(null);
  const articleData = data.content;

  const handleScroll = () => {
    const topPos = (heroImageWrap.current.getBoundingClientRect().top + 50);
    if (topPos < 0 && isScrollTop === false) {
      setIsScrollTop(true);
    } else if (topPos > 0 && isScrollTop === true) {
      setIsScrollTop(false);
    }
  };

  useEffect(() => {
    window.addEventListener('scroll', handleScroll);
    return () => {
      window.removeEventListener('scroll', handleScroll);
    };
  });

  return (
    <HeroImageWrap ref={heroImageWrap}>
      <ImageWrap>
        <ResponsiveImage smallImg={articleData.images.small} largeImg={articleData.images.large} />
      </ImageWrap>
      <Overlay className={isScrollTop ? 'fade-out' : 'fade-in'} />
      <TextOverlay>
        <div className={isScrollTop ? 'fade-out' : 'fade-in'}>
          <span>{articleData.text.articleType}</span>
          <h1>{articleData.text.articleName}</h1>
          <h3><i>with</i></h3>
          <h3>{articleData.text.featuring}</h3>
        </div>
      </TextOverlay>
    </HeroImageWrap>
  );
};

HeroImageWithText.propTypes = {
  data: PropTypes.shape({
    content: PropTypes.shape({
      images: PropTypes.shape({
        small: PropTypes.string.isRequired,
        large: PropTypes.string.isRequired
      }).isRequired,
      text: PropTypes.shape({
        articleType: PropTypes.string.isRequired,
        articleName: PropTypes.string.isRequired,
        featuring: PropTypes.string.isRequired
      }).isRequired
    })
  }).isRequired
};

export default HeroImageWithText;
