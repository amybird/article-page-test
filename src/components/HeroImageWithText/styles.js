import styled from 'styled-components';
import mq from '../../styles/mediaQueries';

export const HeroImageWrap = styled.div`
  position: relative;

  .fade-in {
    visibility: visible;
    opacity: 1;
    transition: opacity 0.3s linear;
  }

  .fade-out {
    visibility: hidden;
    opacity: 0;
    transition: visibility 0s 0.3s, opacity 0.3s linear;
  }

  ${mq.medium`
    height: 500px;
    overflow: hidden;
  `}
`;

export const Overlay = styled.div`
  background-color: rgba(0, 0, 0, 0.6);
  width: 100%;
  height: 100%;
  position: absolute;
  top: 0;
`;

export const ImageWrap = styled.div`
  ${mq.medium`
    position: absolute;
  `}
`;

export const TextOverlay = styled.div`
  position: absolute;
  color: white;
  width: 100%;
  top: 50%;
  padding: 0 30px;
  * {
    font-weight: lighter;
  }

  span {
    text-transform: uppercase;
    letter-spacing: 1px;
    font-size: 10px;
  }

  h1 {
    margin: 0;
    text-transform: uppercase;
    font-family: serif;
    font-size: 30px;
  }

  h3 {
    margin: 0;
    text-align: right;
    font-family: serif;
  }

  ${mq.medium`
    padding: 0 0 0;
    div {
      max-width: 1080px;
      margin: 0 auto;
      height: 100%;
      /* padding: 250px 0 0; */
    }
    span {
      font-size: 16px;
    }
    h1 {
      font-size: 70px;
      letter-spacing: 2px;
    }
    h3 {
      font-size: 26px;
    }
  `}
`;
