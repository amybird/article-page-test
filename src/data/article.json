{
    "articleName": "Standing strong",
    "articleType": "Cover Story",
    "featuring": "Danai Gurira",
    "content": [
        {
            "contentType":"HeroImage",
            "content":{
                "images":{
                    "small":"https://cache.net-a-porter.com/content/images/story-head-content-SUBBED-1556275864012.jpeg/w600_q65.jpeg",
                    "large": "https://cache.net-a-porter.com/content/images/story-head-content-SUBBED-1556275864012.jpeg/w1900_q65.jpeg"
                },
                "text":{
                    "articleType": "Cover Story",
                    "articleName": "Standing strong",
                    "featuring": "Danai Gurira"
                }
            }
        },
        {
            "contentType":"StickyImageWithText",
            "content":{
                "imageLeft": false,
                "paragraphs": [
                    "Going unnoticed is not in Danai Gurira’s repertoire. When the 41-year-old actress walks into the roof restaurant of the Dream hotel in Hollywood, wearing faded denim, a cream silk blouse and mirrored aviators, she practically gleams in the morning light. It’s hard to say if the other patrons recognize her as General Okoye, the fierce protector of the Wakanda royal family in Black Panther, or the equally imposing zombie-slayer Michonne from The Walking Dead, or are just struck by the way she carries herself, but all eyes are on her.",
                    "“I’m running a writers’ room near here,” Gurira says, sitting down and peering over the railing at the sprawl of surrounding offices and studios. She is talking about the newest line on her résumé: head writer for a television mini-series. Aside from her commanding performances on screen, she’s also a feted writer whose plays have garnered multiple awards, including a Tony nomination. Now she’s adapting Chimamanda Ngozi Adichie’s acclaimed novel Americanah – a love story following a Nigerian couple through their complicated lives – with her longtime friend Lupita Nyong’o. The two became close in 2015 when Nyong’o starred in Gurira’s all-female play Eclipsed, about the Liberian civil war."  
                ],
                "quote": "“It’s about collaborating NOT cooperating. I’m pretty willing to have a CLEAR conversation about what I think WORKS, and what doesn’t, and why”",
                "image":"https://cache.net-a-porter.com/content/images/story-body-content-SUBBED-0-1556266075382.jpeg/w950_q65.jpeg"
            }
        },
        {
            "contentType":"StickyImageWithText",
            "content":{
                "imageLeft": true,
                "paragraphs": [
                    "Gurira is tight-lipped about her upcoming departure from The Walking Dead, for which she’s scheduled to guest star in a handful of episodes in the tenth season. “We start shooting soon, so I’ll have to start fitting in my workouts,” she says of her Facetime sessions with her trainer. In a month, she’ll return to Atlanta, now practically her third home – as the location for the long-running zombie series as well as the Marvel shoots – alongside her apartment in Brooklyn and home in Hollywood. “I have a really beautiful family in The Walking Dead,” is what she tells me. “We have pot luck dinners. We go to each other’s homes. We check in with each other. I also talk a lot to my show runner, Scott Gimple, who is kind of my ‘Bat phone’ as I navigate running a writer’s room.”",
                    "For the task of adapting Americanah, Gurira is actually uniquely qualified. “Funnily, me and Chimamanda have these interesting parallels,” she says. “We were both raised by academics, either near or on an African university campus. We saw our countries go through various things when we were growing up. And we came back to America for school. Lupita, too.”",
                    "Born to African parents in the US, Gurira returned to her family’s native Zimbabwe when she was five, and then came back to the States for college in Minnesota. “I am definitely a strong meshing of both worlds,” she says, explaining that she feels American and Zimbabwean in equal parts. She first ‘noticed’ her cultural composition when she was a student at Macalester College in Saint Paul, Minnesota. “I knew when I got here that I had to be vigilant. Being in a land of such access and excess, you have to really be clear about what you’re doing and why, and how you’re going to be responsible with all of the opportunities available to you as an American citizen. And it was clear to me that I was going to retain a tangible and active connection to Zimbabwe. When I moved to Africa as a girl, I thought I was American. Then when I came back, I realized how African I was.”"  
                ],
                "quote": "“I am a strong meshing of BOTH worlds. When I moved to Africa as a girl, I thought I was AMERICAN. Then when I came back, I realized how AFRICAN I was”",
                "image":"https://cache.net-a-porter.com/content/images/story-body-content-SUBBED-0-1556276906399.jpeg/w950_q65.jpeg"
            }
        },
        {
            "contentType":"Paragraph",
            "content":{
                "paragraphs": [
                    "Going unnoticed is not in Danai Gurira’s repertoire. When the 41-year-old actress walks into the roof restaurant of the Dream hotel in Hollywood, wearing faded denim, a cream silk blouse and mirrored aviators, she practically gleams in the morning light. It’s hard to say if the other patrons recognize her as General Okoye, the fierce protector of the Wakanda royal family in Black Panther, or the equally imposing zombie-slayer Michonne from The Walking Dead, or are just struck by the way she carries herself, but all eyes are on her.",
                    "“I’m running a writers’ room near here,” Gurira says, sitting down and peering over the railing at the sprawl of surrounding offices and studios. She is talking about the newest line on her résumé: head writer for a television mini-series. Aside from her commanding performances on screen, she’s also a feted writer whose plays have garnered multiple awards, including a Tony nomination. Now she’s adapting Chimamanda Ngozi Adichie’s acclaimed novel Americanah – a love story following a Nigerian couple through their complicated lives – with her longtime friend Lupita Nyong’o. The two became close in 2015 when Nyong’o starred in Gurira’s all-female play Eclipsed, about the Liberian civil war."  
                ]
            }
        }
    ]
}
