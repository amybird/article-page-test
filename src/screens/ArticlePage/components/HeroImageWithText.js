import React, {Component} from 'react';
import ResponsiveImage from '../../../ui/ResponsiveImage';
import styled, {css} from 'styled-components/macro';
import mq from '../../../styles/mediaQueries';

class HeroImage extends Component {
    state = { 
        isScrollTop: false
    }
    heroImageWrap = React.createRef();
    parallaxText = React.createRef();
    image = React.createRef();

    componentDidMount(){
        window.addEventListener('scroll', this.handleScroll)
    }

    componentWillUnmount(){
        window.removeEventListener('scroll', this.handleScroll)
    }

    parallax() {        
        const parallaxEl = window.innerWidth < 767 ? this.parallaxText.current : this.image.current;
        const yPos = (window.innerWidth < 767 ? 0 : -50) - window.pageYOffset/(window.innerWidth < 767 ? 40 : 8);	
        parallaxEl.style.top = 50 + yPos + "%";
    }

    handleScroll = () => { 
        const topPos = this.heroImageWrap.current.getBoundingClientRect().top;

        if (topPos < 1 && this.state.isScrollTop === false){
            this.setState({
                isScrollTop: true
            })
        } else if (topPos > 0 && this.state.isScrollTop === true) { 
            this.setState({
                isScrollTop: false
            })
        }

        this.parallax()
    }

    render() {
        const data = this.props[0];
        const { isScrollTop } = this.state;

        return (
            <HeroImageWrap ref={this.heroImageWrap}>
                <ImageWrap ref={this.image}>
                    <ResponsiveImage smallImg={data.images.small} largeImg={data.images.large} />
                </ImageWrap>
                <Overlay className={isScrollTop ? 'fade-out' : 'fade-in'}/>
                <TextOverlay ref={this.parallaxText} >
                    <div className={isScrollTop ? 'fade-out' : 'fade-in'}>
                        <span>{data.text.articleType}</span>
                        <h1>{data.text.articleName}</h1>
                        <h3><i>with</i></h3>
                        <h3>{data.text.featuring}</h3>
                    </div>
                </TextOverlay>
            </HeroImageWrap>
        )
    }
}

const HeroImageWrap = styled.div`
    position:relative;

    .fade-in { 
        visibility: visible;
        opacity: 1;
        transition: opacity .3s linear;
    }

    .fade-out { 
        visibility: hidden;
        opacity: 0;
        transition: visibility 0s 0.3s, opacity 0.3s linear;
    }

    ${mq.medium(css`
        height:500px;
        overflow:hidden;
    `)}
`;

const Overlay = styled.div`
    background-color: rgba(0, 0, 0, 0.6);
    width: 100%;
    height: 100%;
    position:absolute;
    top:0;
`;

const ImageWrap = styled.div`
    ${mq.medium(css`
        position:absolute;
    `)}
`;

const TextOverlay = styled.div`
    position: absolute; 
    color: white;
    width: 100%;
    top:50%;
    padding:0 30px;
    * {
        font-weight:lighter;
    }

    span { 
        text-transform: uppercase;
        letter-spacing:1px;
        font-size: 10px;
    }

    h1 { 
        margin:0;
        text-transform:uppercase;
        font-family:serif;
        font-size: 30px;
    }

    h3 { 
        margin:0;
        text-align:right;
        font-family:serif;
    }

    ${mq.medium(css`
        padding:0  0 0;
        div { 
            max-width:1080px;
            margin:0 auto;
            height: 100%;
            /* padding: 250px 0 0; */
        }
        span { 
            font-size:16px;
        }
        h1 { 
            font-size: 70px;
            letter-spacing: 2px;
        }
        h3 { 
            font-size:26px;
        }

    `)}

`;



export default HeroImage;