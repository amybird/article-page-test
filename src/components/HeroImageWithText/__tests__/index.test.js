import React from 'react';
import renderer from 'react-test-renderer';
import HeroImageWithText from '../index';
import { data } from '../../../data/article';

describe('Container', () => {
  it('should render a container', () => {
    const tree = renderer
      .create(<HeroImageWithText data={data.content[0]} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
