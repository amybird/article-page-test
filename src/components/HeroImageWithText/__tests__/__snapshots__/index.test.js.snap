// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container should render a container 1`] = `
.c0 {
  position: relative;
}

.c0 .fade-in {
  visibility: visible;
  opacity: 1;
  -webkit-transition: opacity 0.3s linear;
  transition: opacity 0.3s linear;
}

.c0 .fade-out {
  visibility: hidden;
  opacity: 0;
  -webkit-transition: visibility 0s 0.3s,opacity 0.3s linear;
  transition: visibility 0s 0.3s,opacity 0.3s linear;
}

.c2 {
  background-color: rgba(0,0,0,0.6);
  width: 100%;
  height: 100%;
  position: absolute;
  top: 0;
}

.c3 {
  position: absolute;
  color: white;
  width: 100%;
  top: 50%;
  padding: 0 30px;
}

.c3 * {
  font-weight: lighter;
}

.c3 span {
  text-transform: uppercase;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  font-size: 10px;
}

.c3 h1 {
  margin: 0;
  text-transform: uppercase;
  font-family: serif;
  font-size: 30px;
}

.c3 h3 {
  margin: 0;
  text-align: right;
  font-family: serif;
}

@media (min-width:768px) {
  .c0 {
    height: 500px;
    overflow: hidden;
  }
}

@media (min-width:768px) {
  .c1 {
    position: absolute;
  }
}

@media (min-width:768px) {
  .c3 {
    padding: 0 0 0;
  }

  .c3 div {
    max-width: 1080px;
    margin: 0 auto;
    height: 100%;
  }

  .c3 span {
    font-size: 16px;
  }

  .c3 h1 {
    font-size: 70px;
    -webkit-letter-spacing: 2px;
    -moz-letter-spacing: 2px;
    -ms-letter-spacing: 2px;
    letter-spacing: 2px;
  }

  .c3 h3 {
    font-size: 26px;
  }
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <picture>
      <source
        media="(min-width: 768px)"
        srcSet="https://cache.net-a-porter.com/content/images/story-head-content-SUBBED-1556275864012.jpeg/w1900_q65.jpeg"
      />
      <img
        alt="this is a lovely image"
        src="https://cache.net-a-porter.com/content/images/story-head-content-SUBBED-1556275864012.jpeg/w600_q65.jpeg"
      />
    </picture>
  </div>
  <div
    className="c2 fade-in"
  />
  <div
    className="c3"
  >
    <div
      className="fade-in"
    >
      <span>
        Cover Story
      </span>
      <h1>
        Standing strong
      </h1>
      <h3>
        <i>
          with
        </i>
      </h3>
      <h3>
        Danai Gurira
      </h3>
    </div>
  </div>
</div>
`;
